### PR TITLE
Update rubocop-govuk to 4.0.0.pre.1

### PR DESCRIPTION
**Pull request by kevindew**
Trello: https://trello.com/c/pqXbKZiC/215-test-rubocop-govuk-400pre1-prior-to-400-release

This bumps to the 4.0.0 pre release of rubocop-govuk which is part of [trialling the new rubocop-govuk release](https://github.com/alphagov/rubocop-govuk/issues/129). It seems to have gone into this app with some annoyances in the 
`Style/CommentAnnotation` cop and `Lint/MissingSuper` catching a few oddities.

As part of this I've also made the Gemfile consistent with GOV.UK Rails app conventions

### CODE REVIEW

File: lib/debug/rank_eval.rb [hunk 1]
```diff
@@ -53,8 +53,8 @@ def evaluate
 
       {
         score: result["metric_score"],
-        query_scores: result["details"].each_with_object({}) do |(query, detail), acc|
-          acc[query] = detail["metric_score"]
+        query_scores: result["details"].transform_values do |detail|
```
Comment (leenagupte) [inline lib/debug/rank_eval.rb:56]: This is much nicer, but I can't believe I've never seen it used before.

File: lib/indexer/comparer.rb [hunk 2]
```diff
@@ -21,8 +21,8 @@ def run
       log "processing: #{@old_index_name}/#{@new_index_name}"
 
       search_body = {}
-      if @filtered_format
-        search_body[:post_filter] = { term: { format: @filtered_format } } if @filtered_format
+      if @filtered_format && @filtered_format
```
Comment (leenagupte) [inline lib/indexer/comparer.rb]: This is checking `@filtered_format` twice, so I think the second clause can be removed.

File: lib/tasks/export.rake [hunk 3]
```diff
@@ -23,9 +23,10 @@ namespace :export do
         csv << fields.map do |f|
           value = hit["_source"][f]
 
-          if value.is_a? Hash
+          case value
+          when Hash
```
Comment (leenagupte) [inline lib/tasks/export.rake:27]: This is nice!

File: lib/tasks/debug.rake [hunk 4]
```diff
@@ -75,7 +75,7 @@ namespace :debug do
 
       maxlen = results[:query_scores].map { |query, _| query.length }.max
       results[:query_scores].each do |query, score|
-        puts "#{(query + ':').ljust(maxlen + 1)} #{score}"
+        puts "#{("#{query}:").ljust(maxlen + 1)} #{score}"
```
Comment (leenagupte) [inline lib/tasks/debug.rake]: This is a bit of nitpick, but the nested `"` look a bit odd. Can they be replaced with single-quotes?
```suggestion
        puts "#{('#{query}:').ljust(maxlen + 1)} #{score}"
```

File: lib/indexer/comparer.rb [hunk 5]
```diff
@@ -21,7 +21,7 @@ def run
       log "processing: #{@old_index_name}/#{@new_index_name}"
 
       search_body = {}
-      if @filtered_format && @filtered_format
+      if @filtered_format
```
Comment (leenagupte) [inline lib/indexer/comparer.rb:24]: Ah! you caught it!

File: lib/search/escaping.rb [hunk 6]
```diff
@@ -1,8 +1,10 @@
 module Search
   module Escaping
-    LUCENE_SPECIAL_CHARACTERS = Regexp.new("(" + %w[
-      + - && || ! ( ) { } [ ] ^ " ~ * ? : \\ /
-    ].map { |s| Regexp.escape(s) }.join("|") + ")")
+    LUCENE_SPECIAL_CHARACTERS = begin
+      chars = %w[+ - && || ! ( ) { } [ ] ^ " ~ * ? : \\ /]
+      escaped = chars.map { |s| Regexp.escape(s) }.join("|")
+      Regexp.new("(#{escaped})")
+    end
```
Comment (leenagupte) [inline lib/search/escaping.rb:7]: This is much much nicer!

**Review by leenagupte — APPROVED** [review]
There's one comment about nested double-quotes, but it's not a blocker.  LGTM 👍

File: lib/tasks/debug.rake (hunk 4 above)
Comment (kevindew) [reply lib/tasks/debug.rake]: Thanks for the suggestion here, I'd agree that this looks a bit odd. 

I felt that the underlying code was already rather cryptic so I've gone with an approach of breaking that out instead in https://github.com/alphagov/search-api/pull/2266/commits/97f56af45cbc3e27796279ded2ed4132f4e904e1

Comment (leenagupte) [reply lib/tasks/debug.rake]: Splitting them out is better. The approval still stands 👍

**Pull request merged**